Gems/Atom/RPI/Code/Source/RPI.Public/Pass/CopyPass.cpp

```diff
@@ -138,11 +294,323 @@ namespace AZ
+        {
+            RHI::CopyItemType copyType = GetCopyItemType();
+            auto inputId = GetInputBinding(0).GetAttachment()->GetAttachmentId();
+            switch (copyType)
+            {
+            case AZ::RHI::CopyItemType::Image:
+                [[fallthrough]];
+            case AZ::RHI::CopyItemType::ImageToBuffer:
+                {
+                    // copy image to read back buffer since only buffer can be accessed by host
+                    const auto* sourceImage = context.GetImage(inputId);
+                    if (!sourceImage)
+                    {
+                        AZ_Warning("AttachmentReadback", false, "Failed to find attachment image %s for copy to buffer", inputId.GetCStr());
+                        return;
+                    }
+                    const auto& sourceImageDescriptor = sourceImage->GetDescriptor();
+                    const uint16_t sourceMipSlice = m_data.m_imageSourceSubresource.m_mipSlice;
+                    RHI::ImageSubresourceRange sourceRange(sourceMipSlice, sourceMipSlice, 0, 0);
+                    sourceRange.m_aspectFlags = RHI::ImageAspectFlags::Color;
+
+                    RHI::ImageAspect sourceImageAspect = RHI::ImageAspect::Color;
+                    RHI::ImageAspectFlags sourceImageAspectFlags = RHI::GetImageAspectFlags(sourceImageDescriptor.m_format);
+                    if (RHI::CheckBitsAll(sourceImageAspectFlags, RHI::ImageAspectFlags::Depth))
+                    {
+                        sourceImageAspect = RHI::ImageAspect::Depth;
+                        sourceRange.m_aspectFlags = RHI::ImageAspectFlags::Depth;
+                    }
+
+                    AZStd::vector<RHI::SingleDeviceImageSubresourceLayout> sourceImageSubResourcesLayouts;
+                    sourceImageSubResourcesLayouts.resize_no_construct(sourceImageDescriptor.m_mipLevels);
+                    size_t sourceTotalSizeInBytes = 0;
+                    sourceImage->GetDeviceImage(m_data.m_sourceDeviceIndex)
+                        ->GetSubresourceLayouts(sourceRange, sourceImageSubResourcesLayouts.data(), &sourceTotalSizeInBytes);
+                    AZ::u64 sourceByteCount = sourceTotalSizeInBytes;
+
+                    RPI::CommonBufferDescriptor desc;
+                    desc.m_poolType = RPI::CommonBufferPoolType::ReadBack;
+                    desc.m_bufferName = AZStd::string(GetPathName().GetStringView()) + "_hostbuffer";
+                    desc.m_byteCount = sourceByteCount;
+                    m_device1HostBuffer[m_currentBufferIndex] = BufferSystemInterface::Get()->CreateBufferFromCommonPool(desc);
+                    desc.m_bufferName = AZStd::string(GetPathName().GetStringView()) + "_hostbuffer2";
+                    desc.m_poolType = RPI::CommonBufferPoolType::Staging;
+                    m_device2HostBuffer[m_currentBufferIndex] = BufferSystemInterface::Get()->CreateBufferFromCommonPool(desc);
```

Comment:
If we are doing copies each frame won't we be allocating new memory each frame instead of re-using memory in a ring buffer like fashion?